### DISCUSSION TITLE
Recommended approach for updating the embded dashboard theme programmatically

### DISCUSSION THREAD

**Discussion by troypoulter** [plausible/analytics#1893] [Anything Else]
I've read through [Embed your dashboard on another website using an iFrame](https://plausible.io/docs/embed-dashboard) and outlines well how to embed your dashboard and choose one of the three themes: `light`, `dark` or `system`.

In my site, I have a dark/light mode toggle which can override whatever the use has in their `system` setting, and I'd like it to also update this dashboard to use the relevant theme.

What I've tried (see script at bottom) is whenever the toggle is selected, listen for this and update the `iframe` `src` attribute to have the updated `theme` value.

Now this works and the `src` is updated, and it reloads, but I get this error:

`Failed to execute 'postMessage' on 'DOMWindow': The target origin provided ('https://plausible.io') does not match the recipient window's origin ('http://localhost:1313'). embed.host.js:1`.

I get this error is thrown by `https://plausible.io/js/embed.host.js` so it comes back to, what would be the recommended approach for wanting to update the theme of the embedded dashboard programmatically, so we don't get the above error but still follow your recommended pattern in the docs?

---

**Entire code to add the `iframe` along with my custom `script`**

```javascript
<iframe plausible-embed src="https://plausible.io/share/example.com?auth=xxx&embed=true&theme=system&background=transparent" scrolling="no" frameborder="0" loading="lazy" style="width: 1px; min-width: 100%; height: 1600px;"></iframe>
<div style="font-size: 14px; padding-bottom: 14px;">Stats powered by <a target="_blank" style="color: #4F46E5; text-decoration: underline;" href="https://plausible.io">Plausible Analytics</a></div>
<script async src="https://plausible.io/js/embed.host.js"></script>
<script>
    function setPlausibleTheme(theme) {
        let plausible = document.querySelector("iframe");
        if (plausible) {
            plausible.src = "https://plausible.io/share/example.com?auth=xxx&embed=true&theme=" + theme + "&background=transparent";
        }
    }

    window.addEventListener('onColorSchemeChange', (e) => {
        setPlausibleTheme(e.detail)
    });
</script>
```

_Note: Have removed my URL to keep the script above generic._

**Comment by RobertJoonas** [comment]
Hi @troypoulter!

I tried reproducing the issue but couldn't get the same error. It actually worked all nicely for me with a simple button to toggle between the themes and a very similar code to what you wrote here.

There's this Stack Overflow thread on the same issue with several reasons for the same error message. Perhaps you're able to find the answer here: https://stackoverflow.com/questions/22194409/failed-to-execute-postmessage-on-domwindow-the-target-origin-provided-does.

Hope this helps!

**Reply by troypoulter** [reply]
Apologies for the delay @RobertJoonas!

It turns out that I was reading the error wrong, the error was coming from `embed.host.js` and not this script.

Thanks for sharing that SO link, was good to understand how all this works a bit better anyway!
